{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "LzuNmQ_jk9Q9"
      },
      "source": [
        "### MEDC0106: Bioinformatics in Applied Biomedical Science\n",
        "\n",
        "<p align=\"center\">\n",
        "  <img src=\"https://github.com/MEDC0106/PythonWorkshop/blob/main/resources/static/Banner.png?raw=1\" alt=\"MEDC0106 Banner\" width=\"90%\"/>\n",
        "  <br>\n",
        "</p>\n",
        "\n",
        "---------------------------------------------------------------\n",
        "\n",
        "# 05 - Handling sequences with Biopython\n",
        "\n",
        "*Written by:* Mateusz Kaczyński\n",
        "\n",
        "**This notebook introduces the Biopython library, focusing on in-depth sequence handling—the cornerstone of bioinformatics.**\n",
        "\n",
        "\n",
        "### What is Biopython?\n",
        "\n",
        "**Biopython** is a widely-used, open-source toolkit for computational biology and bioinformatics. It includes tools, resource connectors, and functions for executing standard bioinformatics tasks and parsing common data formats.\n",
        "\n",
        "*Parsing is just a fancier word for interpreting or analyzing data to extract meaningful information.*\n",
        "\n",
        "### Why Biopython?\n",
        "\n",
        "**Biopython** is the de facto standard for accessing numerous bioinformatics databases and tools, making it easier to share your work and results. It simplifies interaction with libraries written in other languages and technologies, as well as those hosted online. While it is possible to use these resources directly, Biopython streamlines the process and provides the essential components you need.\n",
        "\n",
        "-----\n",
        "\n",
        "## Contents\n",
        "\n",
        "1. [Sequence basics](#Sequence-basics)\n",
        "2. [Transcription and translation](#Transcription-and-translation)\n",
        "3. [Alignment](#Alignment)\n",
        "4. [Downloading and reading FASTA files](#Downloading-and-reading-fasta-files)\n",
        "5. [Discussion](#Discussion)\n",
        "\n",
        "-----\n",
        "\n",
        "### Extra resources:\n",
        "\n",
        "- [Official Biopython tutorial](http://biopython.org/DIST/docs/tutorial/Tutorial.html) - A comprehensive guide to the capabilities of the library.\n",
        "- [Biopython API documentation](https://biopython.org/docs/latest/api/index.html) - A long, detailed list of all methods and connectors provided by Biopython.\n",
        "- [Rosalind](http://rosalind.info) - A bioinformatics learning platform that includes exercises.\n",
        "\n",
        "-----"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "collapsed": false,
        "jupyter": {
          "outputs_hidden": false
        },
        "id": "b94ZVMKnk9Q-"
      },
      "source": [
        "### Installing Biopython\n",
        "In order to use the library locally, we first need to install it. Most Python installations come with `pip` tool that can be run directly from the notebook. It will download and install relevant packages from [Python Package Index website (PyPI)](https://pypi.org)  "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {
        "jupyter": {
          "outputs_hidden": false
        },
        "id": "AgA70yfnk9Q-",
        "outputId": "a2c30bd0-cc65-48ad-d40a-bc1628c5bd68",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting Biopython\n",
            "  Downloading biopython-1.86-cp312-cp312-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl.metadata (13 kB)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.12/dist-packages (from Biopython) (2.0.2)\n",
            "Downloading biopython-1.86-cp312-cp312-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl (3.2 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.2/3.2 MB\u001b[0m \u001b[31m39.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: Biopython\n",
            "Successfully installed Biopython-1.86\n"
          ]
        }
      ],
      "source": [
        "!pip install Biopython"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "9IhO6_PTk9Q_"
      },
      "source": [
        "### Importing Biopython\n",
        "When installed, **Biopython** can be included in the project by importing it as the `Bio` module.\n",
        "\n",
        "It is not uncommon for the library to be called differently than the imported name. In this case, **Biopython** is referenced as **Bio**."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {
        "id": "YCGq5nQuk9Q_",
        "outputId": "2b34656b-aa9c-4b2b-ecfe-c78da0850ced",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Module Bio version 1.86\n"
          ]
        }
      ],
      "source": [
        "import Bio\n",
        "print(\"Module\", Bio.__name__, \"version\", Bio.__version__)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_rioqGRWk9Q_"
      },
      "source": [
        "-----\n",
        "## Sequence basics\n",
        "Managing DNA, RNA, or peptide sequences is central to any bioinformatics workflow. The ability to understand, analyse, and transform these sequences is one of the most fundamental applications of computing in biology.\n",
        "\n",
        "Here’s how to define a basic sequence."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {
        "id": "BTilxrWyk9Q_",
        "outputId": "d0175f01-2710-4b39-a011-aa83b9c6e9a4",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "Seq('ACTG')"
            ]
          },
          "metadata": {},
          "execution_count": 6
        }
      ],
      "source": [
        "from Bio.Seq import Seq\n",
        "dna_sequence = Seq(\"ACTG\")\n",
        "dna_sequence"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "lyvbJfb9k9Q_"
      },
      "source": [
        "`Seq` is a submodule within `Bio`, specifically designed to handle biological sequences. The `Seq` class represents a sequence of nucleotides or amino acids and provides various methods for analysing and manipulating these sequences.\n",
        "\n",
        "You can see that a sequence can be created from a basic string that represents the ordered elements–in this case, a DNA sequence. However, it is no longer treated as a simple data type.\n",
        "\n",
        "Let’s examine this more closely with Python’s representation."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {
        "id": "1wNihAMrk9RA",
        "outputId": "62671354-acc7-41be-d5d0-76d41c17b53b",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "The sequence compared with the original string: True\n",
            "The type of dna_sequence: <class 'Bio.Seq.Seq'>\n"
          ]
        }
      ],
      "source": [
        "print(\"The sequence compared with the original string:\", dna_sequence == \"ACTG\")\n",
        "print(\"The type of dna_sequence:\", type(dna_sequence))"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "uVFS9lKGk9RA"
      },
      "source": [
        "The first line checks if `dna_sequence` is equivalent to the original string `\"ACTG\"`, demonstrating that the content is the same even though the data type is different.\n",
        "\n",
        "The second line displays the type of `dna_sequence`, which will show it is a `Bio.Seq.Seq` object rather than a standard Python string, highlighting the enhanced functionality provided by the `Seq` class."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "kE45rHgKk9RA"
      },
      "source": [
        "#### Objects and classes\n",
        "We can see that, while `dna_sequence` can be compared with the original string without any issues, its type is actually `Seq`. This means that `dna_sequence` is an object of the `Seq` class.\n",
        "    \n",
        "Let's unpack these key concepts:\n",
        "  - **Object** - A fundamental building block in programming that can hold data (properties) and functions (methods). In Python, almost everything is an object, from numbers to complex data types.\n",
        "  - **Class** - A blueprint or template for creating new objects. A class allows us to group data and related functions together, defining how each instance (object) should behave.\n",
        "  - `Seq` - A particular class defined in the `Bio.Seq` module within Biopython, designed to handle biological sequences.\n",
        "\n",
        "When in doubt, we can use the `help()` function to understand how to use `Seq`. This will display detailed information about the class, including available methods, usage examples, and descriptions of its functionalities."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {
        "id": "WbRBw2WFk9RA",
        "outputId": "40be9ad7-1923-470c-93b0-e75441576ccf",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Help on Seq in module Bio.Seq object:\n",
            "\n",
            "class Seq(_SeqAbstractBaseClass)\n",
            " |  Seq(data: str | bytes | bytearray | Bio.Seq._SeqAbstractBaseClass | Bio.Seq.SequenceDataAbstractBaseClass | dict | None, length: int | None = None)\n",
            " |\n",
            " |  Read-only sequence object (essentially a string with biological methods).\n",
            " |\n",
            " |  Like normal python strings, our basic sequence object is immutable.\n",
            " |  This prevents you from doing my_seq[5] = \"A\" for example, but does allow\n",
            " |  Seq objects to be used as dictionary keys.\n",
            " |\n",
            " |  The Seq object provides a number of string like methods (such as count,\n",
            " |  find, split and strip).\n",
            " |\n",
            " |  The Seq object also provides some biological methods, such as complement,\n",
            " |  reverse_complement, transcribe, back_transcribe and translate (which are\n",
            " |  not applicable to protein sequences).\n",
            " |\n",
            " |  Method resolution order:\n",
            " |      Seq\n",
            " |      _SeqAbstractBaseClass\n",
            " |      abc.ABC\n",
            " |      builtins.object\n",
            " |\n",
            " |  Methods defined here:\n",
            " |\n",
            " |  __hash__(self)\n",
            " |      Hash of the sequence as a string for comparison.\n",
            " |\n",
            " |      See Seq object comparison documentation (method ``__eq__`` in\n",
            " |      particular) as this has changed in Biopython 1.65. Older versions\n",
            " |      would hash on object identity.\n",
            " |\n",
            " |  __init__(self, data: str | bytes | bytearray | Bio.Seq._SeqAbstractBaseClass | Bio.Seq.SequenceDataAbstractBaseClass | dict | None, length: int | None = None)\n",
            " |      Create a Seq object.\n",
            " |\n",
            " |      Arguments:\n",
            " |       - data - Sequence, required (string)\n",
            " |       - length - Sequence length, used only if data is None or a dictionary (integer)\n",
            " |\n",
            " |      You will typically use Bio.SeqIO to read in sequences from files as\n",
            " |      SeqRecord objects, whose sequence will be exposed as a Seq object via\n",
            " |      the seq property.\n",
            " |\n",
            " |      However, you can also create a Seq object directly:\n",
            " |\n",
            " |      >>> from Bio.Seq import Seq\n",
            " |      >>> my_seq = Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\")\n",
            " |      >>> my_seq\n",
            " |      Seq('MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF')\n",
            " |      >>> print(my_seq)\n",
            " |      MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
            " |\n",
            " |      To create a Seq object with for a sequence of known length but\n",
            " |      unknown sequence contents, use None for the data argument and pass\n",
            " |      the sequence length for the length argument. Trying to access the\n",
            " |      sequence contents of a Seq object created in this way will raise\n",
            " |      an UndefinedSequenceError:\n",
            " |\n",
            " |      >>> my_undefined_sequence = Seq(None, 20)\n",
            " |      >>> my_undefined_sequence\n",
            " |      Seq(None, length=20)\n",
            " |      >>> len(my_undefined_sequence)\n",
            " |      20\n",
            " |      >>> print(my_undefined_sequence)\n",
            " |      Traceback (most recent call last):\n",
            " |      ...\n",
            " |      Bio.Seq.UndefinedSequenceError: Sequence content is undefined\n",
            " |\n",
            " |      If the sequence contents is known for parts of the sequence only, use\n",
            " |      a dictionary for the data argument to pass the known sequence segments:\n",
            " |\n",
            " |      >>> my_partially_defined_sequence = Seq({3: \"ACGT\"}, 10)\n",
            " |      >>> my_partially_defined_sequence\n",
            " |      Seq({3: 'ACGT'}, length=10)\n",
            " |      >>> len(my_partially_defined_sequence)\n",
            " |      10\n",
            " |      >>> print(my_partially_defined_sequence)\n",
            " |      Traceback (most recent call last):\n",
            " |      ...\n",
            " |      Bio.Seq.UndefinedSequenceError: Sequence content is only partially defined\n",
            " |      >>> my_partially_defined_sequence[3:7]\n",
            " |      Seq('ACGT')\n",
            " |      >>> print(my_partially_defined_sequence[3:7])\n",
            " |      ACGT\n",
            " |\n",
            " |  ----------------------------------------------------------------------\n",
            " |  Data descriptors defined here:\n",
            " |\n",
            " |  __dict__\n",
            " |      dictionary for instance variables\n",
            " |\n",
            " |  __weakref__\n",
            " |      list of weak references to the object\n",
            " |\n",
            " |  ----------------------------------------------------------------------\n",
            " |  Data and other attributes defined here:\n",
            " |\n",
            " |  __abstractmethods__ = frozenset()\n",
            " |\n",
            " |  __annotations__ = {'_data': bytes | Bio.Seq.SequenceDataAbstractBaseCl...\n",
            " |\n",
            " |  ----------------------------------------------------------------------\n",
            " |  Methods inherited from _SeqAbstractBaseClass:\n",
            " |\n",
            " |  __add__(self, other)\n",
            " |      Add a sequence or string to this sequence.\n",
            " |\n",
            " |      >>> from Bio.Seq import Seq, MutableSeq\n",
            " |      >>> Seq(\"MELKI\") + \"LV\"\n",
            " |      Seq('MELKILV')\n",
            " |      >>> MutableSeq(\"MELKI\") + \"LV\"\n",
            " |      MutableSeq('MELKILV')\n",
            " |\n",
            " |  __bytes__(self)\n",
            " |\n",
            " |  __contains__(self, item)\n",
            " |      Return True if item is a subsequence of the sequence, and False otherwise.\n",
            " |\n",
            " |      e.g.\n",
            " |\n",
            " |      >>> from Bio.Seq import Seq, MutableSeq\n",
            " |      >>> my_dna = Seq(\"ATATGAAATTTGAAAA\")\n",
            " |      >>> \"AAA\" in my_dna\n",
            " |      True\n",
            " |      >>> Seq(\"AAA\") in my_dna\n",
            " |      True\n",
            " |      >>> MutableSeq(\"AAA\") in my_dna\n",
            " |      True\n",
            " |\n",
            " |  __eq__(self, other)\n",
            " |      Compare the sequence to another sequence or a string.\n",
            " |\n",
            " |      Sequences are equal to each other if their sequence contents is\n",
            " |      identical:\n",
            " |\n",
            " |      >>> from Bio.Seq import Seq, MutableSeq\n",
            " |      >>> seq1 = Seq(\"ACGT\")\n",
            " |      >>> seq2 = Seq(\"ACGT\")\n",
            " |      >>> mutable_seq = MutableSeq(\"ACGT\")\n",
            " |      >>> seq1 == seq2\n",
            " |      True\n",
            " |      >>> seq1 == mutable_seq\n",
            " |      True\n",
            " |      >>> seq1 == \"ACGT\"\n",
            " |      True\n",
            " |\n",
            " |      Note that the sequence objects themselves are not identical to each\n",
            " |      other:\n",
            " |\n",
            " |      >>> id(seq1) == id(seq2)\n",
            " |      False\n",
            " |      >>> seq1 is seq2\n",
            " |      False\n",
            " |\n",
            " |      Sequences can also be compared to strings, ``bytes``, and ``bytearray``\n",
            " |      objects:\n",
            " |\n",
            " |      >>> seq1 == \"ACGT\"\n",
            " |      True\n",
            " |      >>> seq1 == b\"ACGT\"\n",
            " |      True\n",
            " |      >>> seq1 == bytearray(b\"ACGT\")\n",
            " |      True\n",
            " |\n",
            " |  __ge__(self, other)\n",
            " |      Implement the greater-than or equal operand.\n",
            " |\n",
            " |  __getitem__(self, index)\n",
            " |      Return a subsequence as a single letter or as a sequence object.\n",
            " |\n",
            " |      If the index is an integer, a single letter is returned as a Python\n",
            " |      string:\n",
            " |\n",
            " |      >>> seq = Seq('ACTCGACGTCG')\n",
            " |      >>> seq[5]\n",
            " |      'A'\n",
            " |\n",
            " |      Otherwise, a new sequence object of the same class is returned:\n",
            " |\n",
            " |      >>> seq[5:8]\n",
            " |      Seq('ACG')\n",
            " |      >>> mutable_seq = MutableSeq('ACTCGACGTCG')\n",
            " |      >>> mutable_seq[5:8]\n",
            " |      MutableSeq('ACG')\n",
            " |\n",
            " |  __gt__(self, other)\n",
            " |      Implement the greater-than operand.\n",
            " |\n",
            " |  __imul__(self, other)\n",
            " |      Multiply the sequence object by other and assign.\n",
            " |\n",
            " |      >>> from Bio.Seq import Seq\n",
            " |      >>> seq = Seq('ATG')\n",
            " |      >>> seq *= 2\n",
            " |      >>> seq\n",
            " |      Seq('ATGATG')\n",
            " |\n",
            " |      Note that this is different from in-place multiplication. The ``seq``\n",
            " |      variable is reassigned to the multiplication result, but any variable\n",
            " |      pointing to ``seq`` will remain unchanged:\n",
            " |\n",
            " |      >>> seq = Seq('ATG')\n",
            " |      >>> seq2 = seq\n",
            " |      >>> id(seq) == id(seq2)\n",
            " |      True\n",
            " |      >>> seq *= 2\n",
            " |      >>> seq\n",
            " |      Seq('ATGATG')\n",
            " |      >>> seq2\n",
            " |      Seq('ATG')\n",
            " |      >>> id(seq) == id(seq2)\n",
            " |      False\n",
            " |\n",
            " |  __iter__(self)\n",
            " |      Return an iterable of the sequence.\n",
            " |\n",
            " |  __le__(self, other)\n",
            " |      Implement the less-than or equal operand.\n",
            " |\n",
            " |  __len__(self)\n",
            " |      Return the length of the sequence.\n",
            " |\n",
            " |  __lt__(self, other)\n",
            " |      Implement the less-than operand.\n",
            " |\n",
            " |  __mul__(self, other)\n",
            " |      Multiply sequence by integer.\n",
            " |\n",
            " |      >>> from Bio.Seq import Seq, MutableSeq\n",
            " |      >>> Seq('ATG') * 2\n",
            " |      Seq('ATGATG')\n",
            " |      >>> MutableSeq('ATG') * 2\n",
            " |      MutableSeq('ATGATG')\n",
            " |\n",
            " |  __radd__(self, other)\n",
            " |      Add a sequence string on the left.\n",
            " |\n",
            " |      >>> from Bio.Seq import Seq, MutableSeq\n",
            " |      >>> \"LV\" + Seq(\"MELKI\")\n",
            " |      Seq('LVMELKI')\n",
            " |      >>> \"LV\" + MutableSeq(\"MELKI\")\n",
            " |      MutableSeq('LVMELKI')\n",
            " |\n",
            " |      Adding two sequence objects is handled via the __add__ method.\n",
            " |\n",
            " |  __repr__(self)\n",
            " |      Return (truncated) representation of the sequence.\n",
            " |\n",
            " |  __rmul__(self, other)\n",
            " |      Multiply integer by sequence.\n",
            " |\n",
            " |      >>> from Bio.Seq import Seq\n",
            " |      >>> 2 * Seq('ATG')\n",
            " |      Seq('ATGATG')\n",
            " |\n",
            " |  __str__(self)\n",
            " |      Return the full sequence as a python string.\n",
            " |\n",
            " |  back_transcribe(self, inplace=False)\n",
            " |      Return the DNA sequence from an RNA sequence by creating a new Seq object.\n",
            " |\n",
            " |      >>> from Bio.Seq import Seq\n",
            " |      >>> messenger_rna = Seq(\"AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG\")\n",
            " |      >>> messenger_rna\n",
            " |      Seq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
            " |      >>> messenger_rna.back_transcribe()\n",
            " |      Seq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
            " |\n",
            " |      The sequence is modified in-place and returned if inplace is True:\n",
            " |\n",
            " |      >>> sequence = MutableSeq(\"AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG\")\n",
            " |      >>> sequence\n",
            " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
            " |      >>> sequence.back_transcribe()\n",
            " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
            " |      >>> sequence\n",
            " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
            " |\n",
            " |      >>> sequence.back_transcribe(inplace=True)\n",
            " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
            " |      >>> sequence\n",
            " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
            " |\n",
            " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
            " |      ``transcribe`` is called on a ``Seq`` object with ``inplace=True``.\n",
            " |\n",
            " |      Trying to back-transcribe DNA has no effect, If you have a nucleotide\n",
            " |      sequence which might be DNA or RNA (or even a mixture), calling the\n",
            " |      back-transcribe method will ensure any U becomes T.\n",
            " |\n",
            " |      Trying to back-transcribe a protein sequence will replace any U for\n",
            " |      Selenocysteine with T for Threonine, which is biologically meaningless.\n",
            " |\n",
            " |      >>> from Bio.Seq import Seq\n",
            " |      >>> my_protein = Seq(\"MAIVMGRU\")\n",
            " |      >>> my_protein.back_transcribe()\n",
            " |      Seq('MAIVMGRT')\n",
            " |\n",
            " |  complement(self, inplace=False)\n",
            " |      Return the complement as a DNA sequence.\n",
            " |\n",
            " |      >>> Seq(\"CGA\").complement()\n",
            " |      Seq('GCT')\n",
            " |\n",
            " |      Any U in the sequence is treated as a T:\n",
            " |\n",
            " |      >>> Seq(\"CGAUT\").complement()\n",
            " |      Seq('GCTAA')\n",
            " |\n",
            " |      In contrast, ``complement_rna`` returns an RNA sequence:\n",
            " |\n",
            " |      >>> Seq(\"CGAUT\").complement_rna()\n",
            " |      Seq('GCUAA')\n",
            " |\n",
            " |      The sequence is modified in-place and returned if inplace is True:\n",
            " |\n",
            " |      >>> my_seq = MutableSeq(\"CGA\")\n",
            " |      >>> my_seq\n",
            " |      MutableSeq('CGA')\n",
            " |      >>> my_seq.complement()\n",
            " |      MutableSeq('GCT')\n",
            " |      >>> my_seq\n",
            " |      MutableSeq('CGA')\n",
            " |\n",
            " |      >>> my_seq.complement(inplace=True)\n",
            " |      MutableSeq('GCT')\n",
            " |      >>> my_seq\n",
            " |      MutableSeq('GCT')\n",
            " |\n",
            " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
            " |      ``complement_rna`` is called on a ``Seq`` object with ``inplace=True``.\n",
            " |\n",
            " |  complement_rna(self, inplace=False)\n",
            " |      Return the complement as an RNA sequence.\n",
            " |\n",
            " |      >>> Seq(\"CGA\").complement_rna()\n",
            " |      Seq('GCU')\n",
            " |\n",
            " |      Any T in the sequence is treated as a U:\n",
            " |\n",
            " |      >>> Seq(\"CGAUT\").complement_rna()\n",
            " |      Seq('GCUAA')\n",
            " |\n",
            " |      In contrast, ``complement`` returns a DNA sequence by default:\n",
            " |\n",
            " |      >>> Seq(\"CGA\").complement()\n",
            " |      Seq('GCT')\n",
            " |\n",
            " |      The sequence is modified in-place and returned if inplace is True:\n",
            " |\n",
            " |      >>> my_seq = MutableSeq(\"CGA\")\n",
            " |      >>> my_seq\n",
            " |      MutableSeq('CGA')\n",
            " |      >>> my_seq.complement_rna()\n",
            " |      MutableSeq('GCU')\n",
            " |      >>> my_seq\n",
            " |      MutableSeq('CGA')\n",
            " |\n",
            " |      >>> my_seq.complement_rna(inplace=True)\n",
            " |      MutableSeq('GCU')\n",
            " |      >>> my_seq\n",
            " |      MutableSeq('GCU')\n",
            " |\n",
            " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
            " |      ``complement_rna`` is called on a ``Seq`` object with ``inplace=True``.\n",
            " |\n",
            " |  count(self, sub, start=None, end=None)\n",
            " |      Return a non-overlapping count, like that of a python string.\n",
            " |\n",
            " |      The number of occurrences of substring argument sub in the\n",
            " |      (sub)sequence given by [start:end] is returned as an integer.\n",
            " |      Optional arguments start and end are interpreted as in slice\n",
            " |      notation.\n",
            " |\n",
            " |      Arguments:\n",
            " |       - sub - a string or another Seq object to look for\n",
            " |       - start - optional integer, slice start\n",
            " |       - end - optional integer, slice end\n",
            " |\n",
            " |      e.g.\n",
            " |\n",
            " |      >>> from Bio.Seq import Seq\n",
            " |      >>> my_seq = Seq(\"AAAATGA\")\n",
            " |      >>> print(my_seq.count(\"A\"))\n",
            " |      5\n",
            " |      >>> print(my_seq.count(\"ATG\"))\n",
            " |      1\n",
            " |      >>> print(my_seq.count(Seq(\"AT\")))\n",
            " |      1\n",
            " |      >>> print(my_seq.count(\"AT\", 2, -1))\n",
            " |      1\n",
            " |\n",
            " |      HOWEVER, please note because the ``count`` method of Seq and MutableSeq\n",
            " |      objects, like that of Python strings, do a non-overlapping search, this\n",
            " |      may not give the answer you expect:\n",
            " |\n",
            " |      >>> \"AAAA\".count(\"AA\")\n",
            " |      2\n",
            " |      >>> print(Seq(\"AAAA\").count(\"AA\"))\n",
            " |      2\n",
            " |\n",
            " |      For an overlapping search, use the ``count_overlap`` method:\n",
            " |\n",
            " |      >>> print(Seq(\"AAAA\").count_overlap(\"AA\"))\n",
            " |      3\n",
            " |\n",
            " |  count_overlap(self, sub, start=None, end=None)\n",
            " |      Return an overlapping count.\n",
            " |\n",
            " |      Returns an integer, the number of occurrences of substring\n",
            " |      argument sub in the (sub)sequence given by [start:end].\n",
            " |      Optional arguments start and end are interpreted as in slice\n",
            " |      notation.\n",
            " |\n",
            " |      Arguments:\n",
            " |       - sub - a string or another Seq object to look for\n",
            " |       - start - optional integer, slice start\n",
            " |       - end - optional integer, slice end\n",
            " |\n",
            " |      e.g.\n",
            " |\n",
            " |      >>> from Bio.Seq import Seq\n",
            " |      >>> print(Seq(\"AAAA\").count_overlap(\"AA\"))\n",
            " |      3\n",
            " |      >>> print(Seq(\"ATATATATA\").count_overlap(\"ATA\"))\n",
            " |      4\n",
            " |      >>> print(Seq(\"ATATATATA\").count_overlap(\"ATA\", 3, -1))\n",
            " |      1\n",
            " |\n",
            " |      For a non-overlapping search, use the ``count`` method:\n",
            " |\n",
            " |      >>> print(Seq(\"AAAA\").count(\"AA\"))\n",
            " |      2\n",
            " |\n",
            " |      Where substrings do not overlap, ``count_overlap`` behaves the same as\n",
            " |      the ``count`` method:\n",
            " |\n",
            " |      >>> from Bio.Seq import Seq\n",
            " |      >>> my_seq = Seq(\"AAAATGA\")\n",
            " |      >>> print(my_seq.count_overlap(\"A\"))\n",
            " |      5\n",
            " |      >>> my_seq.count_overlap(\"A\") == my_seq.count(\"A\")\n",
            " |      True\n",
            " |      >>> print(my_seq.count_overlap(\"ATG\"))\n",
            " |      1\n",
            " |      >>> my_seq.count_overlap(\"ATG\") == my_seq.count(\"ATG\")\n",
            " |      True\n",
            " |      >>> print(my_seq.count_overlap(Seq(\"AT\")))\n",
            " |      1\n",
            " |      >>> my_seq.count_overlap(Seq(\"AT\")) == my_seq.count(Seq(\"AT\"))\n",
            " |      True\n",
            " |      >>> print(my_seq.count_overlap(\"AT\", 2, -1))\n",
            " |      1\n",
            " |      >>> my_seq.count_overlap(\"AT\", 2, -1) == my_seq.count(\"AT\", 2, -1)\n",
            " |      True\n",
            " |\n",
            " |      HOWEVER, do not use this method for such cases because the\n",
            " |      count() method is much for efficient.\n",
            " |\n",
            " |  endswith(self, suffix, start=None, end=None)\n",
            " |      Return True if the sequence ends with the given suffix, False otherwise.\n",
            " |\n",
            " |      Return True if the sequence ends with the specified suffix\n",
            " |      (a string or another Seq object), False otherwise.\n",
            " |      With optional start, test sequence beginning at that position.\n",
            " |      With optional end, stop comparing sequence at that position.\n",
            " |      suffix can also be a tuple of strings to try.  e.g.\n",
            " |\n",
            " |      >>> from Bio.Seq import Seq\n",
            " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
            " |      >>> my_rna.endswith(\"UUG\")\n",
            " |      True\n",
            " |      >>> my_rna.endswith(\"AUG\")\n",
            " |      False\n",
            " |      >>> my_rna.endswith(\"AUG\", 0, 18)\n",
            " |      True\n",
            " |      >>> my_rna.endswith((\"UCC\", \"UCA\", \"UUG\"))\n",
            " |      True\n",
            " |\n",
            " |  find(self, sub, start=None, end=None)\n",
            " |      Return the lowest index in the sequence where subsequence sub is found.\n",
            " |\n",
            " |      With optional arguments start and end, return the lowest index in the\n",
            " |      sequence such that the subsequence sub is contained within the sequence\n",
            " |      region [start:end].\n",
            " |\n",
            " |      Arguments:\n",
            " |       - sub - a string or another Seq or MutableSeq object to search for\n",
            " |       - start - optional integer, slice start\n",
            " |       - end - optional integer, slice end\n",
            " |\n",
            " |      Returns -1 if the subsequence is NOT found.\n",
            " |\n",
            " |      e.g. Locating the first typical start codon, AUG, in an RNA sequence:\n",
            " |\n",
            " |      >>> from Bio.Seq import Seq\n",
            " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
            " |      >>> my_rna.find(\"AUG\")\n",
            " |      3\n",
            " |\n",
            " |      The next typical start codon can then be found by starting the search\n",
            " |      at position 4:\n",
            " |\n",
            " |      >>> my_rna.find(\"AUG\", 4)\n",
            " |      15\n",
            " |\n",
            " |      See the ``search`` method to find the locations of multiple subsequences\n",
            " |      at the same time.\n",
            " |\n",
            " |  index(self, sub, start=None, end=None)\n",
            " |      Return the lowest index in the sequence where subsequence sub is found.\n",
            " |\n",
            " |      With optional arguments start and end, return the lowest index in the\n",
            " |      sequence such that the subsequence sub is contained within the sequence\n",
            " |      region [start:end].\n",
            " |\n",
            " |      Arguments:\n",
            " |       - sub - a string or another Seq or MutableSeq object to search for\n",
            " |       - start - optional integer, slice start\n",
            " |       - end - optional integer, slice end\n",
            " |\n",
            " |      Raises a ValueError if the subsequence is NOT found.\n",
            " |\n",
            " |      e.g. Locating the first typical start codon, AUG, in an RNA sequence:\n",
            " |\n",
            " |      >>> from Bio.Seq import Seq\n",
            " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
            " |      >>> my_rna.index(\"AUG\")\n",
            " |      3\n",
            " |\n",
            " |      The next typical start codon can then be found by starting the search\n",
            " |      at position 4:\n",
            " |\n",
            " |      >>> my_rna.index(\"AUG\", 4)\n",
            " |      15\n",
            " |\n",
            " |      This method performs the same search as the ``find`` method.  However,\n",
            " |      if the subsequence is not found, ``find`` returns -1 while ``index``\n",
            " |      raises a ValueError:\n",
            " |\n",
            " |      >>> my_rna.index(\"T\")\n",
            " |      Traceback (most recent call last):\n",
            " |                 ...\n",
            " |      ValueError: ...\n",
            " |      >>> my_rna.find(\"T\")\n",
            " |      -1\n",
            " |\n",
            " |      See the ``search`` method to find the locations of multiple subsequences\n",
            " |      at the same time.\n",
            " |\n",
            " |  islower(self)\n",
            " |      Return True if all ASCII characters in data are lowercase.\n",
            " |\n",
            " |      If there are no cased characters, the method returns False.\n",
            " |\n",
            " |  isupper(self)\n",
            " |      Return True if all ASCII characters in data are uppercase.\n",
            " |\n",
            " |      If there are no cased characters, the method returns False.\n",
            " |\n",
            " |  join(self, other)\n",
            " |      Return a merge of the sequences in other, spaced by the sequence from self.\n",
            " |\n",
            " |      Accepts a Seq object, MutableSeq object, or string (and iterates over\n",
            " |      the letters), or an iterable containing Seq, MutableSeq, or string\n",
            " |      objects. These arguments will be concatenated with the calling sequence\n",
            " |      as the spacer:\n",
            " |\n",
            " |      >>> concatenated = Seq('NNNNN').join([Seq(\"AAA\"), Seq(\"TTT\"), Seq(\"PPP\")])\n",
            " |      >>> concatenated\n",
            " |      Seq('AAANNNNNTTTNNNNNPPP')\n",
            " |\n",
            " |      Joining the letters of a single sequence:\n",
            " |\n",
            " |      >>> Seq('NNNNN').join(Seq(\"ACGT\"))\n",
            " |      Seq('ANNNNNCNNNNNGNNNNNT')\n",
            " |      >>> Seq('NNNNN').join(\"ACGT\")\n",
            " |      Seq('ANNNNNCNNNNNGNNNNNT')\n",
            " |\n",
            " |  lower(self, inplace=False)\n",
            " |      Return the sequence in lower case.\n",
            " |\n",
            " |      An lower-case copy of the sequence is returned if inplace is False,\n",
            " |      the default value:\n",
            " |\n",
            " |      >>> from Bio.Seq import Seq, MutableSeq\n",
            " |      >>> my_seq = Seq(\"VHLTPeeK*\")\n",
            " |      >>> my_seq\n",
            " |      Seq('VHLTPeeK*')\n",
            " |      >>> my_seq.lower()\n",
            " |      Seq('vhltpeek*')\n",
            " |      >>> my_seq.upper()\n",
            " |      Seq('VHLTPEEK*')\n",
            " |      >>> my_seq\n",
            " |      Seq('VHLTPeeK*')\n",
            " |\n",
            " |      The sequence is modified in-place and returned if inplace is True:\n",
            " |\n",
            " |      >>> my_seq = MutableSeq(\"VHLTPeeK*\")\n",
            " |      >>> my_seq\n",
            " |      MutableSeq('VHLTPeeK*')\n",
            " |      >>> my_seq.lower()\n",
            " |      MutableSeq('vhltpeek*')\n",
            " |      >>> my_seq.upper()\n",
            " |      MutableSeq('VHLTPEEK*')\n",
            " |      >>> my_seq\n",
            " |      MutableSeq('VHLTPeeK*')\n",
            " |\n",
            " |      >>> my_seq.lower(inplace=True)\n",
            " |      MutableSeq('vhltpeek*')\n",
            " |      >>> my_seq\n",
            " |      MutableSeq('vhltpeek*')\n",
            " |      >>> my_seq.upper(inplace=True)\n",
            " |      MutableSeq('VHLTPEEK*')\n",
            " |      >>> my_seq\n",
            " |      MutableSeq('VHLTPEEK*')\n",
            " |\n",
            " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
            " |      ``lower`` is called on a ``Seq`` object with ``inplace=True``.\n",
            " |\n",
            " |      See also the ``upper`` method.\n",
            " |\n",
            " |  lstrip(self, chars=None, inplace=False)\n",
            " |      Return a sequence object with leading and trailing ends stripped.\n",
            " |\n",
            " |      With default arguments, leading whitespace is removed:\n",
            " |\n",
            " |      >>> seq = Seq(\" ACGT \")\n",
            " |      >>> seq.lstrip()\n",
            " |      Seq('ACGT ')\n",
            " |      >>> seq\n",
            " |      Seq(' ACGT ')\n",
            " |\n",
            " |      If ``chars`` is given and not ``None``, remove characters in ``chars``\n",
            " |      from the leading end instead.  The order of the characters to be removed\n",
            " |      is not important:\n",
            " |\n",
            " |      >>> Seq(\"ACGACGTTACG\").lstrip(\"GCA\")\n",
            " |      Seq('TTACG')\n",
            " |\n",
            " |      A copy of the sequence is returned if ``inplace`` is ``False`` (the\n",
            " |      default value).  If ``inplace`` is ``True``, the sequence is stripped\n",
            " |      in-place and returned.\n",
            " |\n",
            " |      >>> seq = MutableSeq(\" ACGT \")\n",
            " |      >>> seq.lstrip()\n",
            " |      MutableSeq('ACGT ')\n",
            " |      >>> seq\n",
            " |      MutableSeq(' ACGT ')\n",
            " |      >>> seq.lstrip(inplace=True)\n",
            " |      MutableSeq('ACGT ')\n",
            " |      >>> seq\n",
            " |      MutableSeq('ACGT ')\n",
            " |\n",
            " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
            " |      ``lstrip`` is called on a ``Seq`` object with ``inplace=True``.\n",
            " |\n",
            " |      See also the strip and rstrip methods.\n",
            " |\n",
            " |  removeprefix(self, prefix, inplace=False)\n",
            " |      Return a new Seq object with prefix (left) removed.\n",
            " |\n",
            " |      This behaves like the python string method of the same name.\n",
            " |\n",
            " |      e.g. Removing a start Codon:\n",
            " |\n",
            " |      >>> from Bio.Seq import Seq\n",
            " |      >>> my_seq = Seq(\"ATGGTGTGTGT\")\n",
            " |      >>> my_seq\n",
            " |      Seq('ATGGTGTGTGT')\n",
            " |      >>> my_seq.removeprefix('ATG')\n",
            " |      Seq('GTGTGTGT')\n",
            " |\n",
            " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
            " |      ``removeprefix`` is called on a ``Seq`` object with ``inplace=True``.\n",
            " |\n",
            " |      See also the removesuffix method.\n",
            " |\n",
            " |  removesuffix(self, suffix, inplace=False)\n",
            " |      Return a new Seq object with suffix (right) removed.\n",
            " |\n",
            " |      This behaves like the python string method of the same name.\n",
            " |\n",
            " |      e.g. Removing a stop codon:\n",
            " |\n",
            " |      >>> from Bio.Seq import Seq\n",
            " |      >>> my_seq = Seq(\"GTGTGTGTTAG\")\n",
            " |      >>> my_seq\n",
            " |      Seq('GTGTGTGTTAG')\n",
            " |      >>> stop_codon = Seq(\"TAG\")\n",
            " |      >>> my_seq.removesuffix(stop_codon)\n",
            " |      Seq('GTGTGTGT')\n",
            " |\n",
            " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
            " |      ``removesuffix`` is called on a ``Seq`` object with ``inplace=True``.\n",
            " |\n",
            " |      See also the removeprefix method.\n",
            " |\n",
            " |  replace(self, old, new, inplace=False)\n",
            " |      Return a copy with all occurrences of subsequence old replaced by new.\n",
            " |\n",
            " |      >>> s = Seq(\"ACGTAACCGGTT\")\n",
            " |      >>> t = s.replace(\"AC\", \"XYZ\")\n",
            " |      >>> s\n",
            " |      Seq('ACGTAACCGGTT')\n",
            " |      >>> t\n",
            " |      Seq('XYZGTAXYZCGGTT')\n",
            " |\n",
            " |      For mutable sequences, passing inplace=True will modify the sequence in place:\n",
            " |\n",
            " |      >>> m = MutableSeq(\"ACGTAACCGGTT\")\n",
            " |      >>> t = m.replace(\"AC\", \"XYZ\")\n",
            " |      >>> m\n",
            " |      MutableSeq('ACGTAACCGGTT')\n",
            " |      >>> t\n",
            " |      MutableSeq('XYZGTAXYZCGGTT')\n",
            " |\n",
            " |      >>> m = MutableSeq(\"ACGTAACCGGTT\")\n",
            " |      >>> t = m.replace(\"AC\", \"XYZ\", inplace=True)\n",
            " |      >>> m\n",
            " |      MutableSeq('XYZGTAXYZCGGTT')\n",
            " |      >>> t\n",
            " |      MutableSeq('XYZGTAXYZCGGTT')\n",
            " |\n",
            " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
            " |      ``replace`` is called on a ``Seq`` object with ``inplace=True``.\n",
            " |\n",
            " |  reverse_complement(self, inplace=False)\n",
            " |      Return the reverse complement as a DNA sequence.\n",
            " |\n",
            " |      >>> Seq(\"CGA\").reverse_complement()\n",
            " |      Seq('TCG')\n",
            " |\n",
            " |      Any U in the sequence is treated as a T:\n",
            " |\n",
            " |      >>> Seq(\"CGAUT\").reverse_complement()\n",
            " |      Seq('AATCG')\n",
            " |\n",
            " |      In contrast, ``reverse_complement_rna`` returns an RNA sequence:\n",
            " |\n",
            " |      >>> Seq(\"CGA\").reverse_complement_rna()\n",
            " |      Seq('UCG')\n",
            " |\n",
            " |      The sequence is modified in-place and returned if inplace is True:\n",
            " |\n",
            " |      >>> my_seq = MutableSeq(\"CGA\")\n",
            " |      >>> my_seq\n",
            " |      MutableSeq('CGA')\n",
            " |      >>> my_seq.reverse_complement()\n",
            " |      MutableSeq('TCG')\n",
            " |      >>> my_seq\n",
            " |      MutableSeq('CGA')\n",
            " |\n",
            " |      >>> my_seq.reverse_complement(inplace=True)\n",
            " |      MutableSeq('TCG')\n",
            " |      >>> my_seq\n",
            " |      MutableSeq('TCG')\n",
            " |\n",
            " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
            " |      ``reverse_complement`` is called on a ``Seq`` object with\n",
            " |      ``inplace=True``.\n",
            " |\n",
            " |  reverse_complement_rna(self, inplace=False)\n",
            " |      Return the reverse complement as an RNA sequence.\n",
            " |\n",
            " |      >>> Seq(\"CGA\").reverse_complement_rna()\n",
            " |      Seq('UCG')\n",
            " |\n",
            " |      Any T in the sequence is treated as a U:\n",
            " |\n",
            " |      >>> Seq(\"CGAUT\").reverse_complement_rna()\n",
            " |      Seq('AAUCG')\n",
            " |\n",
            " |      In contrast, ``reverse_complement`` returns a DNA sequence:\n",
            " |\n",
            " |      >>> Seq(\"CGA\").reverse_complement()\n",
            " |      Seq('TCG')\n",
            " |\n",
            " |      The sequence is modified in-place and returned if inplace is True:\n",
            " |\n",
            " |      >>> my_seq = MutableSeq(\"CGA\")\n",
            " |      >>> my_seq\n",
            " |      MutableSeq('CGA')\n",
            " |      >>> my_seq.reverse_complement_rna()\n",
            " |      MutableSeq('UCG')\n",
            " |      >>> my_seq\n",
            " |      MutableSeq('CGA')\n",
            " |\n",
            " |      >>> my_seq.reverse_complement_rna(inplace=True)\n",
            " |      MutableSeq('UCG')\n",
            " |      >>> my_seq\n",
            " |      MutableSeq('UCG')\n",
            " |\n",
            " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
            " |      ``reverse_complement_rna`` is called on a ``Seq`` object with\n",
            " |      ``inplace=True``.\n",
            " |\n",
            " |  rfind(self, sub, start=None, end=None)\n",
            " |      Return the highest index in the sequence where subsequence sub is found.\n",
            " |\n",
            " |      With optional arguments start and end, return the highest index in the\n",
            " |      sequence such that the subsequence sub is contained within the sequence\n",
            " |      region [start:end].\n",
            " |\n",
            " |      Arguments:\n",
            " |       - sub - a string or another Seq or MutableSeq object to search for\n",
            " |       - start - optional integer, slice start\n",
            " |       - end - optional integer, slice end\n",
            " |\n",
            " |      Returns -1 if the subsequence is NOT found.\n",
            " |\n",
            " |      e.g. Locating the last typical start codon, AUG, in an RNA sequence:\n",
            " |\n",
            " |      >>> from Bio.Seq import Seq\n",
            " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
            " |      >>> my_rna.rfind(\"AUG\")\n",
            " |      15\n",
            " |\n",
            " |      The location of the typical start codon before that can be found by\n",
            " |      ending the search at position 15:\n",
            " |\n",
            " |      >>> my_rna.rfind(\"AUG\", end=15)\n",
            " |      3\n",
            " |\n",
            " |      See the ``search`` method to find the locations of multiple subsequences\n",
            " |      at the same time.\n",
            " |\n",
            " |  rindex(self, sub, start=None, end=None)\n",
            " |      Return the highest index in the sequence where subsequence sub is found.\n",
            " |\n",
            " |      With optional arguments start and end, return the highest index in the\n",
            " |      sequence such that the subsequence sub is contained within the sequence\n",
            " |      region [start:end].\n",
            " |\n",
            " |      Arguments:\n",
            " |       - sub - a string or another Seq or MutableSeq object to search for\n",
            " |       - start - optional integer, slice start\n",
            " |       - end - optional integer, slice end\n",
            " |\n",
            " |      Returns -1 if the subsequence is NOT found.\n",
            " |\n",
            " |      e.g. Locating the last typical start codon, AUG, in an RNA sequence:\n",
            " |\n",
            " |      >>> from Bio.Seq import Seq\n",
            " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
            " |      >>> my_rna.rindex(\"AUG\")\n",
            " |      15\n",
            " |\n",
            " |      The location of the typical start codon before that can be found by\n",
            " |      ending the search at position 15:\n",
            " |\n",
            " |      >>> my_rna.rindex(\"AUG\", end=15)\n",
            " |      3\n",
            " |\n",
            " |      This method performs the same search as the ``rfind`` method.  However,\n",
            " |      if the subsequence is not found, ``rfind`` returns -1 which ``rindex``\n",
            " |      raises a ValueError:\n",
            " |\n",
            " |      >>> my_rna.rindex(\"T\")\n",
            " |      Traceback (most recent call last):\n",
            " |                 ...\n",
            " |      ValueError: ...\n",
            " |      >>> my_rna.rfind(\"T\")\n",
            " |      -1\n",
            " |\n",
            " |      See the ``search`` method to find the locations of multiple subsequences\n",
            " |      at the same time.\n",
            " |\n",
            " |  rsplit(self, sep=None, maxsplit=-1)\n",
            " |      Return a list of subsequences by splitting the sequence from the right.\n",
            " |\n",
            " |      Return a list of the subsequences in the sequence (as Seq objects),\n",
            " |      using sep as the delimiter string.  If maxsplit is given, at\n",
            " |      most maxsplit splits are done.  If maxsplit is omitted, all\n",
            " |      splits are made.\n",
            " |\n",
            " |      For consistency with the ``rsplit`` method of Python strings, any\n",
            " |      whitespace (tabs, spaces, newlines) is a separator if sep is None, the\n",
            " |      default value\n",
            " |\n",
            " |      e.g.\n",
            " |\n",
            " |      >>> from Bio.Seq import Seq\n",
            " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
            " |      >>> my_aa = my_rna.translate()\n",
            " |      >>> my_aa\n",
            " |      Seq('VMAIVMGR*KGAR*L')\n",
            " |      >>> for pep in my_aa.rsplit(\"*\"):\n",
            " |      ...     pep\n",
            " |      Seq('VMAIVMGR')\n",
            " |      Seq('KGAR')\n",
            " |      Seq('L')\n",
            " |      >>> for pep in my_aa.rsplit(\"*\", 1):\n",
            " |      ...     pep\n",
            " |      Seq('VMAIVMGR*KGAR')\n",
            " |      Seq('L')\n",
            " |\n",
            " |      See also the split method, which splits the sequence starting from the\n",
            " |      beginning:\n",
            " |\n",
            " |      >>> for pep in my_aa.split(\"*\", 1):\n",
            " |      ...     pep\n",
            " |      Seq('VMAIVMGR')\n",
            " |      Seq('KGAR*L')\n",
            " |\n",
            " |  rstrip(self, chars=None, inplace=False)\n",
            " |      Return a sequence object with trailing ends stripped.\n",
            " |\n",
            " |      With default arguments, trailing whitespace is removed:\n",
            " |\n",
            " |      >>> seq = Seq(\" ACGT \")\n",
            " |      >>> seq.rstrip()\n",
            " |      Seq(' ACGT')\n",
            " |      >>> seq\n",
            " |      Seq(' ACGT ')\n",
            " |\n",
            " |      If ``chars`` is given and not ``None``, remove characters in ``chars``\n",
            " |      from the trailing end instead.  The order of the characters to be\n",
            " |      removed is not important:\n",
            " |\n",
            " |      >>> Seq(\"ACGACGTTACG\").rstrip(\"GCA\")\n",
            " |      Seq('ACGACGTT')\n",
            " |\n",
            " |      A copy of the sequence is returned if ``inplace`` is ``False`` (the\n",
            " |      default value).  If ``inplace`` is ``True``, the sequence is stripped\n",
            " |      in-place and returned.\n",
            " |\n",
            " |      >>> seq = MutableSeq(\" ACGT \")\n",
            " |      >>> seq.rstrip()\n",
            " |      MutableSeq(' ACGT')\n",
            " |      >>> seq\n",
            " |      MutableSeq(' ACGT ')\n",
            " |      >>> seq.rstrip(inplace=True)\n",
            " |      MutableSeq(' ACGT')\n",
            " |      >>> seq\n",
            " |      MutableSeq(' ACGT')\n",
            " |\n",
            " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
            " |      ``rstrip`` is called on a ``Seq`` object with ``inplace=True``.\n",
            " |\n",
            " |      See also the strip and lstrip methods.\n",
            " |\n",
            " |  search(self, subs)\n",
            " |      Search the substrings subs in self and yield the index and substring found.\n",
            " |\n",
            " |      Arguments:\n",
            " |       - subs - a list of strings, Seq, MutableSeq, bytes, or bytearray\n",
            " |         objects containing the substrings to search for.\n",
            " |\n",
            " |      >>> from Bio.Seq import Seq\n",
            " |      >>> dna = Seq(\"GTCATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAGTTG\")\n",
            " |      >>> matches = dna.search([\"CC\", Seq(\"ATTG\"), \"ATTG\", Seq(\"CCC\")])\n",
            " |      >>> for index, substring in matches:\n",
            " |      ...     print(index, substring)\n",
            " |      ...\n",
            " |      7 CC\n",
            " |      9 ATTG\n",
            " |      20 CC\n",
            " |      34 CC\n",
            " |      34 CCC\n",
            " |      35 CC\n",
            " |\n",
            " |  split(self, sep=None, maxsplit=-1)\n",
            " |      Return a list of subsequences when splitting the sequence by separator sep.\n",
            " |\n",
            " |      Return a list of the subsequences in the sequence (as Seq objects),\n",
            " |      using sep as the delimiter string.  If maxsplit is given, at\n",
            " |      most maxsplit splits are done.  If maxsplit is omitted, all\n",
            " |      splits are made.\n",
            " |\n",
            " |      For consistency with the ``split`` method of Python strings, any\n",
            " |      whitespace (tabs, spaces, newlines) is a separator if sep is None, the\n",
            " |      default value\n",
            " |\n",
            " |      e.g.\n",
            " |\n",
            " |      >>> from Bio.Seq import Seq\n",
            " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
            " |      >>> my_aa = my_rna.translate()\n",
            " |      >>> my_aa\n",
            " |      Seq('VMAIVMGR*KGAR*L')\n",
            " |      >>> for pep in my_aa.split(\"*\"):\n",
            " |      ...     pep\n",
            " |      Seq('VMAIVMGR')\n",
            " |      Seq('KGAR')\n",
            " |      Seq('L')\n",
            " |      >>> for pep in my_aa.split(\"*\", 1):\n",
            " |      ...     pep\n",
            " |      Seq('VMAIVMGR')\n",
            " |      Seq('KGAR*L')\n",
            " |\n",
            " |      See also the rsplit method, which splits the sequence starting from the\n",
            " |      end:\n",
            " |\n",
            " |      >>> for pep in my_aa.rsplit(\"*\", 1):\n",
            " |      ...     pep\n",
            " |      Seq('VMAIVMGR*KGAR')\n",
            " |      Seq('L')\n",
            " |\n",
            " |  startswith(self, prefix, start=None, end=None)\n",
            " |      Return True if the sequence starts with the given prefix, False otherwise.\n",
            " |\n",
            " |      Return True if the sequence starts with the specified prefix\n",
            " |      (a string or another Seq object), False otherwise.\n",
            " |      With optional start, test sequence beginning at that position.\n",
            " |      With optional end, stop comparing sequence at that position.\n",
            " |      prefix can also be a tuple of strings to try.  e.g.\n",
            " |\n",
            " |      >>> from Bio.Seq import Seq\n",
            " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
            " |      >>> my_rna.startswith(\"GUC\")\n",
            " |      True\n",
            " |      >>> my_rna.startswith(\"AUG\")\n",
            " |      False\n",
            " |      >>> my_rna.startswith(\"AUG\", 3)\n",
            " |      True\n",
            " |      >>> my_rna.startswith((\"UCC\", \"UCA\", \"UCG\"), 1)\n",
            " |      True\n",
            " |\n",
            " |  strip(self, chars=None, inplace=False)\n",
            " |      Return a sequence object with leading and trailing ends stripped.\n",
            " |\n",
            " |      With default arguments, leading and trailing whitespace is removed:\n",
            " |\n",
            " |      >>> seq = Seq(\" ACGT \")\n",
            " |      >>> seq.strip()\n",
            " |      Seq('ACGT')\n",
            " |      >>> seq\n",
            " |      Seq(' ACGT ')\n",
            " |\n",
            " |      If ``chars`` is given and not ``None``, remove characters in ``chars``\n",
            " |      instead.  The order of the characters to be removed is not important:\n",
            " |\n",
            " |      >>> Seq(\"ACGTACGT\").strip(\"TGCA\")\n",
            " |      Seq('')\n",
            " |\n",
            " |      A copy of the sequence is returned if ``inplace`` is ``False`` (the\n",
            " |      default value).  If ``inplace`` is ``True``, the sequence is stripped\n",
            " |      in-place and returned.\n",
            " |\n",
            " |      >>> seq = MutableSeq(\" ACGT \")\n",
            " |      >>> seq.strip()\n",
            " |      MutableSeq('ACGT')\n",
            " |      >>> seq\n",
            " |      MutableSeq(' ACGT ')\n",
            " |      >>> seq.strip(inplace=True)\n",
            " |      MutableSeq('ACGT')\n",
            " |      >>> seq\n",
            " |      MutableSeq('ACGT')\n",
            " |\n",
            " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if ``strip``\n",
            " |      is called on a ``Seq`` object with ``inplace=True``.\n",
            " |\n",
            " |      See also the lstrip and rstrip methods.\n",
            " |\n",
            " |  transcribe(self, inplace=False)\n",
            " |      Transcribe a DNA sequence into RNA and return the RNA sequence as a new Seq object.\n",
            " |\n",
            " |      Following the usual convention, the sequence is interpreted as the\n",
            " |      coding strand of the DNA double helix, not the template strand. This\n",
            " |      means we can get the RNA sequence just by switching T to U.\n",
            " |\n",
            " |      >>> from Bio.Seq import Seq\n",
            " |      >>> coding_dna = Seq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\")\n",
            " |      >>> coding_dna\n",
            " |      Seq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
            " |      >>> coding_dna.transcribe()\n",
            " |      Seq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
            " |\n",
            " |      The sequence is modified in-place and returned if inplace is True:\n",
            " |\n",
            " |      >>> sequence = MutableSeq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\")\n",
            " |      >>> sequence\n",
            " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
            " |      >>> sequence.transcribe()\n",
            " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
            " |      >>> sequence\n",
            " |      MutableSeq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
            " |\n",
            " |      >>> sequence.transcribe(inplace=True)\n",
            " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
            " |      >>> sequence\n",
            " |      MutableSeq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG')\n",
            " |\n",
            " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
            " |      ``transcribe`` is called on a ``Seq`` object with ``inplace=True``.\n",
            " |\n",
            " |      Trying to transcribe an RNA sequence has no effect.\n",
            " |      If you have a nucleotide sequence which might be DNA or RNA\n",
            " |      (or even a mixture), calling the transcribe method will ensure\n",
            " |      any T becomes U.\n",
            " |\n",
            " |      Trying to transcribe a protein sequence will replace any\n",
            " |      T for Threonine with U for Selenocysteine, which has no\n",
            " |      biologically plausible rational.\n",
            " |\n",
            " |      >>> from Bio.Seq import Seq\n",
            " |      >>> my_protein = Seq(\"MAIVMGRT\")\n",
            " |      >>> my_protein.transcribe()\n",
            " |      Seq('MAIVMGRU')\n",
            " |\n",
            " |  translate(self, table='Standard', stop_symbol='*', to_stop=False, cds=False, gap='-')\n",
            " |      Turn a nucleotide sequence into a protein sequence by creating a new sequence object.\n",
            " |\n",
            " |      This method will translate DNA or RNA sequences. It should not\n",
            " |      be used on protein sequences as any result will be biologically\n",
            " |      meaningless.\n",
            " |\n",
            " |      Arguments:\n",
            " |       - table - Which codon table to use?  This can be either a name\n",
            " |         (string), an NCBI identifier (integer), or a CodonTable\n",
            " |         object (useful for non-standard genetic codes).  This\n",
            " |         defaults to the \"Standard\" table.\n",
            " |       - stop_symbol - Single character string, what to use for\n",
            " |         terminators.  This defaults to the asterisk, \"*\".\n",
            " |       - to_stop - Boolean, defaults to False meaning do a full\n",
            " |         translation continuing on past any stop codons (translated as the\n",
            " |         specified stop_symbol).  If True, translation is terminated at\n",
            " |         the first in frame stop codon (and the stop_symbol is not\n",
            " |         appended to the returned protein sequence).\n",
            " |       - cds - Boolean, indicates this is a complete CDS.  If True,\n",
            " |         this checks the sequence starts with a valid alternative start\n",
            " |         codon (which will be translated as methionine, M), that the\n",
            " |         sequence length is a multiple of three, and that there is a\n",
            " |         single in frame stop codon at the end (this will be excluded\n",
            " |         from the protein sequence, regardless of the to_stop option).\n",
            " |         If these tests fail, an exception is raised.\n",
            " |       - gap - Single character string to denote symbol used for gaps.\n",
            " |         Defaults to the minus sign.\n",
            " |\n",
            " |      A ``Seq`` object is returned if ``translate`` is called on a ``Seq``\n",
            " |      object; a ``MutableSeq`` object is returned if ``translate`` is called\n",
            " |      pn a ``MutableSeq`` object.\n",
            " |\n",
            " |      e.g. Using the standard table:\n",
            " |\n",
            " |      >>> coding_dna = Seq(\"GTGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\")\n",
            " |      >>> coding_dna.translate()\n",
            " |      Seq('VAIVMGR*KGAR*')\n",
            " |      >>> coding_dna.translate(stop_symbol=\"@\")\n",
            " |      Seq('VAIVMGR@KGAR@')\n",
            " |      >>> coding_dna.translate(to_stop=True)\n",
            " |      Seq('VAIVMGR')\n",
            " |\n",
            " |      Now using NCBI table 2, where TGA is not a stop codon:\n",
            " |\n",
            " |      >>> coding_dna.translate(table=2)\n",
            " |      Seq('VAIVMGRWKGAR*')\n",
            " |      >>> coding_dna.translate(table=2, to_stop=True)\n",
            " |      Seq('VAIVMGRWKGAR')\n",
            " |\n",
            " |      In fact, GTG is an alternative start codon under NCBI table 2, meaning\n",
            " |      this sequence could be a complete CDS:\n",
            " |\n",
            " |      >>> coding_dna.translate(table=2, cds=True)\n",
            " |      Seq('MAIVMGRWKGAR')\n",
            " |\n",
            " |      It isn't a valid CDS under NCBI table 1, due to both the start codon\n",
            " |      and also the in frame stop codons:\n",
            " |\n",
            " |      >>> coding_dna.translate(table=1, cds=True)\n",
            " |      Traceback (most recent call last):\n",
            " |          ...\n",
            " |      Bio.Data.CodonTable.TranslationError: First codon 'GTG' is not a start codon\n",
            " |\n",
            " |      If the sequence has no in-frame stop codon, then the to_stop argument\n",
            " |      has no effect:\n",
            " |\n",
            " |      >>> coding_dna2 = Seq(\"TTGGCCATTGTAATGGGCCGC\")\n",
            " |      >>> coding_dna2.translate()\n",
            " |      Seq('LAIVMGR')\n",
            " |      >>> coding_dna2.translate(to_stop=True)\n",
            " |      Seq('LAIVMGR')\n",
            " |\n",
            " |      NOTE - Ambiguous codons like \"TAN\" or \"NNN\" could be an amino acid\n",
            " |      or a stop codon.  These are translated as \"X\".  Any invalid codon\n",
            " |      (e.g. \"TA?\" or \"T-A\") will throw a TranslationError.\n",
            " |\n",
            " |      NOTE - This does NOT behave like the python string's translate\n",
            " |      method.  For that use str(my_seq).translate(...) instead\n",
            " |\n",
            " |  upper(self, inplace=False)\n",
            " |      Return the sequence in upper case.\n",
            " |\n",
            " |      An upper-case copy of the sequence is returned if inplace is False,\n",
            " |      the default value:\n",
            " |\n",
            " |      >>> from Bio.Seq import Seq, MutableSeq\n",
            " |      >>> my_seq = Seq(\"VHLTPeeK*\")\n",
            " |      >>> my_seq\n",
            " |      Seq('VHLTPeeK*')\n",
            " |      >>> my_seq.lower()\n",
            " |      Seq('vhltpeek*')\n",
            " |      >>> my_seq.upper()\n",
            " |      Seq('VHLTPEEK*')\n",
            " |      >>> my_seq\n",
            " |      Seq('VHLTPeeK*')\n",
            " |\n",
            " |      The sequence is modified in-place and returned if inplace is True:\n",
            " |\n",
            " |      >>> my_seq = MutableSeq(\"VHLTPeeK*\")\n",
            " |      >>> my_seq\n",
            " |      MutableSeq('VHLTPeeK*')\n",
            " |      >>> my_seq.lower()\n",
            " |      MutableSeq('vhltpeek*')\n",
            " |      >>> my_seq.upper()\n",
            " |      MutableSeq('VHLTPEEK*')\n",
            " |      >>> my_seq\n",
            " |      MutableSeq('VHLTPeeK*')\n",
            " |\n",
            " |      >>> my_seq.lower(inplace=True)\n",
            " |      MutableSeq('vhltpeek*')\n",
            " |      >>> my_seq\n",
            " |      MutableSeq('vhltpeek*')\n",
            " |      >>> my_seq.upper(inplace=True)\n",
            " |      MutableSeq('VHLTPEEK*')\n",
            " |      >>> my_seq\n",
            " |      MutableSeq('VHLTPEEK*')\n",
            " |\n",
            " |      As ``Seq`` objects are immutable, a ``TypeError`` is raised if\n",
            " |      ``upper`` is called on a ``Seq`` object with ``inplace=True``.\n",
            " |\n",
            " |      See also the ``lower`` method.\n",
            " |\n",
            " |  ----------------------------------------------------------------------\n",
            " |  Readonly properties inherited from _SeqAbstractBaseClass:\n",
            " |\n",
            " |  defined\n",
            " |      Return True if the sequence is defined, False if undefined or partially defined.\n",
            " |\n",
            " |      Zero-length sequences are always considered to be defined.\n",
            " |\n",
            " |  defined_ranges\n",
            " |      Return a tuple of the ranges where the sequence contents is defined.\n",
            " |\n",
            " |      The return value has the format ((start1, end1), (start2, end2), ...).\n",
            " |\n",
            " |  ----------------------------------------------------------------------\n",
            " |  Data and other attributes inherited from _SeqAbstractBaseClass:\n",
            " |\n",
            " |  __array_ufunc__ = None\n",
            "\n"
          ]
        }
      ],
      "source": [
        "help(dna_sequence)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "R5y8KUGUk9RA"
      },
      "source": [
        "#### String-like behaviour\n",
        "\n",
        "In the first session, we explored how Python strings (simple objects representing text) behave.\n",
        "\n",
        "The `Seq` class is designed to mimic these string methods, making it easier to interchange information between strings and sequences while providing a familiar interface for the user.\n",
        "\n",
        "Let’s verify this by using some common string methods, including indexing and slicing."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {
        "id": "FTwJvbeKk9RA",
        "outputId": "e818bfe3-1ecf-4442-e0d6-967ac77eca0b",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Length of the sentence:    4\n",
            "AC found in the sequence?: True\n",
            "Sequence ends with 'TG'?:  True\n",
            "The first base:            A\n",
            "The last base:             G\n",
            "First two bases:           AC\n",
            "Adenines in the sequence:  1\n"
          ]
        }
      ],
      "source": [
        "print(\"Length of the sentence:   \", len(dna_sequence))\n",
        "print(\"AC found in the sequence?:\", \"AC\" in dna_sequence)\n",
        "print(\"Sequence ends with 'TG'?: \", dna_sequence.endswith(\"TG\"))\n",
        "print(\"The first base:           \", dna_sequence[0])\n",
        "print(\"The last base:            \", dna_sequence[-1])\n",
        "print(\"First two bases:          \", dna_sequence[0:2])\n",
        "print(\"Adenines in the sequence: \", dna_sequence.count(\"A\"))"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0pTMp8mhk9RA"
      },
      "source": [
        "Does case sensitivity matter when representing bases in a sequence, like `\"A\"` versus `\"a\"` for adenine?\n",
        "\n",
        "Yes, it does. The `Seq` class is case-sensitive, meaning that uppercase and lowercase letters are treated as different characters. It is best to convert all bases to either uppercase or lowercase before performing comparisons or analyses. `Seq` objects provide methods like `.upper()` and `.lower()` to make this easy."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {
        "id": "NXE0YbIPk9RA",
        "outputId": "83526fca-41d7-4751-8beb-b9d09a76808c",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Is the uppercase sequence equal to the mixed case sequence?    False\n",
            "What if we change the original sequence to lowercase?:         False\n",
            "What if we make both sequences the same case? Are they equal?: True\n"
          ]
        }
      ],
      "source": [
        "dna_sequence_mixedcase = Seq(\"acTG\")\n",
        "\n",
        "print(\"Is the uppercase sequence equal to the mixed case sequence?   \",\n",
        "      dna_sequence == dna_sequence_mixedcase)\n",
        "\n",
        "print(\"What if we change the original sequence to lowercase?:        \",\n",
        "      dna_sequence.lower() == dna_sequence_mixedcase)\n",
        "\n",
        "print(\"What if we make both sequences the same case? Are they equal?:\",\n",
        "      dna_sequence.lower() == dna_sequence_mixedcase.lower())"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "e2246msgk9RA"
      },
      "source": [
        "To extend the sequence, we can use the familiar string syntax as `Seq` objects in Biopython behave similarly to Python strings in this context."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {
        "id": "1f7SmgkPk9RA",
        "outputId": "f367cfab-c6bb-4b90-dfae-37b6736f8247",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Extended with another sequence: ACTGTAA\n",
            "Extended with a string        : ACTGTGG\n"
          ]
        }
      ],
      "source": [
        "print(\"Extended with another sequence:\", dna_sequence + Seq(\"TAA\"))\n",
        "print(\"Extended with a string        :\", dna_sequence + \"TGG\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "GieuWQEOk9RB"
      },
      "source": [
        "For the cases where there are gaps, we can remove them using the `.replace()` method."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {
        "id": "yN7TKAxOk9RB",
        "outputId": "cb3c8bc7-9133-4ef1-a79f-4a51ba6211a3",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Sentence with gaps   : ACGT-ACGT-ACGT\n",
            "Sentence without gaps: ACGTACGTACGT\n"
          ]
        }
      ],
      "source": [
        "sequence_with_gaps = Seq(\"ACGT-ACGT-ACGT\")\n",
        "print(\"Sentence with gaps   :\", sequence_with_gaps)\n",
        "print(\"Sentence without gaps:\", sequence_with_gaps.replace(\"-\", \"\"))"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "collapsed": false,
        "jupyter": {
          "outputs_hidden": false
        },
        "id": "5pKpihvUk9RB"
      },
      "source": [
        "#### Sequence analysis\n",
        "\n",
        "**Biopython** offers a variety of tools for sequence analysis. For example, GC content calculation, a common metric in genomics, can be accessed directly from Biopython.\n",
        "[The online documentation](https://biopython.org/docs/dev/api/Bio.SeqUtils.html) provides more details."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {
        "id": "WSfBib5jk9RB",
        "outputId": "5bdd56f5-dfd5-41d6-90cc-c3274523668f",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "The guanine-cytosine content of the sequence: 50.0 %\n"
          ]
        }
      ],
      "source": [
        "from Bio.SeqUtils import gc_fraction\n",
        "print(\"The guanine-cytosine content of the sequence:\", gc_fraction(dna_sequence)*100, \"%\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "uAIwriJsk9RB"
      },
      "source": [
        "## Transcription and translation\n",
        "In addition to their string-like behaviour, `Seq` objects have methods specific to bioinformatics tasks.\n",
        "\n",
        "One of the most powerful features of **Biopython** is its ability to switch between representations, allowing seamless conversion between different sequence types.\n",
        "\n",
        "Let’s explore how to generate complementary and reverse sequences."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {
        "id": "Rb7OxDykk9RB",
        "outputId": "c8b2cc24-7822-4dd0-f74f-7e4cbc9f5656",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Sequence:           ACTG\n",
            "Complement:         TGAC\n",
            "Reverse:            GTCA\n",
            "Reverse complement: CAGT\n"
          ]
        }
      ],
      "source": [
        "print(\"Sequence:          \", dna_sequence)\n",
        "print(\"Complement:        \", dna_sequence.complement())\n",
        "print(\"Reverse:           \", Seq(\"\".join(reversed(dna_sequence))))\n",
        "print(\"Reverse complement:\", dna_sequence.reverse_complement())"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "editable": false,
        "id": "XhrEZZ3rk9RB"
      },
      "source": [
        "Simple DNA-to-RNA transcription can be achieved with the following methods."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {
        "id": "tOmL3DUZk9RB",
        "outputId": "b723dbac-fdda-4ea7-e5b8-a06dfa07ba62",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Sequence:               ACTG\n",
            "Transcribed:            ACUG\n",
            "Reverse complement RNA: CAGU\n"
          ]
        }
      ],
      "source": [
        "print(\"Sequence:              \", dna_sequence)\n",
        "print(\"Transcribed:           \", dna_sequence.transcribe())\n",
        "print(\"Reverse complement RNA:\", dna_sequence.reverse_complement_rna())"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "editable": false,
        "id": "ZpHXFeaTk9RB"
      },
      "source": [
        "The reversal of the process - getting from RNA to DNA is also supported."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {
        "id": "dMRbIFMXk9RB",
        "outputId": "2772cd72-63f2-4c71-c7dd-deb960c06a3a",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Transcribed     : ACUG\n",
            "Back transcribed: ACTG\n"
          ]
        }
      ],
      "source": [
        "transcribed = dna_sequence.transcribe()\n",
        "print(\"Transcribed     :\", transcribed)\n",
        "print(\"Back transcribed:\", transcribed.back_transcribe())"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "editable": false,
        "id": "x40NTdcNk9RB"
      },
      "source": [
        "**Biopython** simplifies the process of translating RNA (or even directly from DNA) into an amino acid sequence.\n",
        "\n",
        "*Note: Different codon tables are also supported, making it flexible for various organisms and use cases.*"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {
        "id": "KKBCyAj_k9RB",
        "outputId": "60376e84-e084-412b-a3ac-7bbc757939bc",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "Seq('TRR')"
            ]
          },
          "metadata": {},
          "execution_count": 14
        }
      ],
      "source": [
        "sequence = Seq(\"ACGCGACGA\")\n",
        "sequence.translate()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "e5kz51R7k9RB"
      },
      "source": [
        "#### Using `Seq` methods directly on strings.\n",
        "It’s possible to bypass the `Seq` class abstraction and directly import and use relevant methods on simple strings."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {
        "id": "UzOA16-ck9RB",
        "outputId": "7a499d98-8527-4afd-cc50-070659823861",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Transcription directly from \"ACGT\" string, without using a Seq object: ACGU\n"
          ]
        }
      ],
      "source": [
        "from Bio.Seq import transcribe\n",
        "print('Transcription directly from \"ACGT\" string, without using a Seq object:', transcribe(\"ACGT\"))"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "NpUoQms7k9RC"
      },
      "source": [
        "## Alignment\n",
        "\n",
        "One of the most common uses of bioinformatics tools is sequence alignment.\n",
        "\n",
        "We will use Biopython’s built-in, out-of-the-box [pairwise aligner](https://biopython.org/docs/latest/api/Bio.Align#Bio.Align.PairwiseAligner), which allows us to align two sequences at a time.\n",
        "\n",
        "We’ll start by importing the `PairwiseAligner`, running a full sequence alignment, and displaying the relevant results."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {
        "id": "fSAdwKQ4k9RC",
        "outputId": "bf55a8e4-1c70-4919-d8db-03de1e918577",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Score: 3.0\n",
            "target            0 ACGT 4\n",
            "                  0 |||. 4\n",
            "query             0 ACGC 4\n",
            "\n"
          ]
        }
      ],
      "source": [
        "from Bio.Align import PairwiseAligner\n",
        "\n",
        "# Initialise the `PairwiseAligner`\n",
        "aligner = PairwiseAligner()\n",
        "\n",
        "# Notice we can provide either `Seq` objects or simple strings as input\n",
        "alignments = aligner.align(Seq(\"ACGT\"), \"ACGC\")\n",
        "\n",
        "# Display alignment results\n",
        "for alignment in alignments:\n",
        "    print(\"Score:\", alignment.score)\n",
        "    print(alignment)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "VIvPnGo7k9RC"
      },
      "source": [
        "The `score` attribute represents the quality of the alignment. By default, it indicates the number of elements that matched exactly. However, both the match score and the penalty for gaps can be adjusted as needed to customize the alignment criteria.\n",
        "\n",
        "**Multiple Sequence Alignment (MSA)** is used when aligning 3 or more sequences simultaneously.\n",
        "\n",
        "For MSA, a list of `SeqRecord` objects is required. The `SeqRecord` class, unlike the basic `Seq` class, allows for additional metadata storage, such as:\n",
        "- **Name**: A short identifier for the sequence.\n",
        "- **Description**: A longer explanation or label.\n",
        "- **Annotations**: Key-value pairs for storing various details.\n",
        "\n",
        "*Note: For MSA, all sequences are expected to be of the same length.*"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {
        "id": "stJOk4qek9RC"
      },
      "outputs": [],
      "source": [
        "from Bio.Align import MultipleSeqAlignment\n",
        "from Bio.SeqRecord import SeqRecord\n",
        "\n",
        "# Define a list of sequences as strings\n",
        "seqs = [\n",
        "    \"ACGTACGT\",\n",
        "    \"ACGTGCGC\",\n",
        "    \"ACGTA--T\",\n",
        "    \"CCGTACGG\",\n",
        "    \"A-GTACCC\",\n",
        "    \"ACGTA--T\",\n",
        "    \"CTG-ACG-\",\n",
        "    \"AGGTACG-\"\n",
        "]\n",
        "\n",
        "# Convert each string in the `seqs` list into a `SeqRecord` object containing a `Seq`\n",
        "seq_records = [SeqRecord(Seq(s)) for s in seqs]\n",
        "\n",
        "# Create a `MultipleSeqAlignment` object with the `SeqRecord` objects\n",
        "aligned = MultipleSeqAlignment(seq_records)\n",
        "\n",
        "# No output is expected for this cell"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "collapsed": false,
        "jupyter": {
          "outputs_hidden": false
        },
        "id": "ghtKQWOKk9RC"
      },
      "source": [
        "We can access the sequence records in the `aligned` object by their index, similar to accessing elements in a `list` or `tuple`."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {
        "jupyter": {
          "outputs_hidden": false
        },
        "id": "s8ls_2wpk9RC",
        "outputId": "cd5b9c27-c865-41bb-bcc3-231d80dc1197",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "ID: <unknown id>\n",
            "Name: <unknown name>\n",
            "Description: <unknown description>\n",
            "Number of features: 0\n",
            "Seq('ACGTACGT') \n",
            "\n",
            "ID: <unknown id>\n",
            "Name: <unknown name>\n",
            "Description: <unknown description>\n",
            "Number of features: 0\n",
            "Seq('ACGTGCGC') \n",
            "\n",
            "ID: <unknown id>\n",
            "Name: <unknown name>\n",
            "Description: <unknown description>\n",
            "Number of features: 0\n",
            "Seq('AGGTACG-')\n"
          ]
        }
      ],
      "source": [
        "# Access the first sequence record\n",
        "print(aligned[0], \"\\n\")\n",
        "\n",
        "# Access the second sequence record\n",
        "print(aligned[1], \"\\n\")\n",
        "\n",
        "# Access the last sequence record\n",
        "print(aligned[-1])"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "collapsed": false,
        "jupyter": {
          "outputs_hidden": false
        },
        "id": "4TAefnl8k9RC"
      },
      "source": [
        "We can also access specific parts of the alignment by slicing both rows and columns, using 2-dimensional slicing."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {
        "jupyter": {
          "outputs_hidden": false
        },
        "id": "L6IjGEGck9RD",
        "outputId": "a50217ec-34e0-4e06-e5d3-6ed855b4b1f4",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "First three sequences:\n",
            " Alignment with 3 rows and 8 columns\n",
            "ACGTACGT <unknown id>\n",
            "ACGTGCGC <unknown id>\n",
            "ACGTA--T <unknown id> \n",
            "\n",
            "First four columns:\n",
            " Alignment with 8 rows and 4 columns\n",
            "ACGT <unknown id>\n",
            "ACGT <unknown id>\n",
            "ACGT <unknown id>\n",
            "CCGT <unknown id>\n",
            "A-GT <unknown id>\n",
            "ACGT <unknown id>\n",
            "CTG- <unknown id>\n",
            "AGGT <unknown id> \n",
            "\n",
            "Rows 1 to 3, columns 2 to 6:\n",
            " Alignment with 3 rows and 4 columns\n",
            "GTGC <unknown id>\n",
            "GTA- <unknown id>\n",
            "GTAC <unknown id>\n"
          ]
        }
      ],
      "source": [
        "# Access the first three sequences in the alignment\n",
        "print(\"First three sequences:\\n\", aligned[:3], \"\\n\")\n",
        "\n",
        "# Access the first four positions (columns) across all sequences\n",
        "print(\"First four columns:\\n\", aligned[:, :4], \"\\n\")\n",
        "\n",
        "# Access rows 1 to 3 and columns 2 to 6\n",
        "print(\"Rows 1 to 3, columns 2 to 6:\\n\", aligned[1:4, 2:6])"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "8mxEUY0nk9RD"
      },
      "source": [
        "Aligning sequences alone provides limited value. To gain meaningful insights, we can extract various statistics from the alignment.\n",
        "\n",
        "Biopython’s `AlignInfo` module historically provided tools such as `SummaryInfo` for examining multiple sequence alignments, for example, extracting columns, computing consensus sequences, or calculating information content.\n",
        "\n",
        "However `SummaryInfo` is now **deprecated**. If you inspect it with `help()`, Biopython clearly indicates that the class is no longer recommended and only exposes a minimal set of older methods."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {
        "id": "ON6raGu-k9RD",
        "outputId": "6bd60667-6836-4a3f-e7e6-12345a1f96aa",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Help on SummaryInfo in module Bio.Align.AlignInfo object:\n",
            "\n",
            "class SummaryInfo(builtins.object)\n",
            " |  SummaryInfo(alignment)\n",
            " |\n",
            " |  Calculate summary info about the alignment.  (DEPRECATED)\n",
            " |\n",
            " |  This class should be used to calculate information summarizing the\n",
            " |  results of an alignment. This may either be straight consensus info\n",
            " |  or more complicated things.\n",
            " |\n",
            " |  Methods defined here:\n",
            " |\n",
            " |  __init__(self, alignment)\n",
            " |      Initialize with the alignment to calculate information on.\n",
            " |\n",
            " |      ic_vector attribute. A list of ic content for each column number.\n",
            " |\n",
            " |  get_column(self, col)\n",
            " |      Return column of alignment.\n",
            " |\n",
            " |  ----------------------------------------------------------------------\n",
            " |  Data descriptors defined here:\n",
            " |\n",
            " |  __dict__\n",
            " |      dictionary for instance variables\n",
            " |\n",
            " |  __weakref__\n",
            " |      list of weak references to the object\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.12/dist-packages/Bio/Align/AlignInfo.py:33: BiopythonDeprecationWarning: The class SummaryInfo has been deprecated. Instead of\n",
            "\n",
            ">>> align_info = AlignInfo.SummaryInfo(msa)\n",
            ">>> sequence = align_info.get_column(1)\n",
            "\n",
            "please use\n",
            "\n",
            ">>> alignment = msa.alignment  # to get a new-style Alignment object\n",
            ">>> sequence = alignment[:, 1]\n",
            "\n",
            "Here, `msa` is a MultipleSeqAlignment object and `alignment` is an\n",
            "`Alignment` object.\n",
            "  warnings.warn(\n"
          ]
        }
      ],
      "source": [
        "from Bio.Align import AlignInfo\n",
        "\n",
        "align_info = AlignInfo.SummaryInfo(aligned)\n",
        "help(align_info)"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Instead, modern Biopython encourages working directly with the new-style `alignment` object. This object allows you to extract alignment statistics using intuitive, NumPy-like slicing."
      ],
      "metadata": {
        "id": "z1fWfWkqpFsa"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 17,
      "metadata": {
        "id": "FtU0Y3ytk9RD",
        "outputId": "5380787b-1e07-46ba-ec48-c00f958d9be2",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 35
        }
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "'CC-CC-CC'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 17
        }
      ],
      "source": [
        "# Get the Alignment object from the MultipleSeqAlignment\n",
        "alignment = aligned.alignment     # aligned is your MultipleSeqAlignment\n",
        "\n",
        "column = alignment[:, 5]   # retrieves column 5\n",
        "column"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "In older versions, `AlignInfo.SummaryInfo()` would have allowed you to compute a “dumb” consensus sequence or inspect per-position scoring matrices directly within Biopython. For example, you could call:\n",
        "\n",
        "```python\n",
        "print(\"The simple consensus:\", align_info.dumb_consensus())\n",
        "print(\"Alignment score for particular bases at a given position:\")\n",
        "print(align_info.pos_specific_score_matrix())\n",
        "```\n",
        "\n",
        "If you want to compute consensus sequences, position-specific statistics, or more sophisticated alignment summaries, the recommended approach today is to use another external library such as [**scikit-bio**](https://scikit.bio/docs/latest/index.html), which provides robust tools for working with multiple sequence alignments.\n",
        "\n",
        "After installing it with `!pip install scikit-bio`, you can run code like the snippet below to generate pairwise alignments, construct MSAs, and compute a majority-rule consensus using maintained and up-to-date functionality.\n",
        "\n",
        "```python\n",
        "from skbio import Protein, TabularMSA\n",
        "from skbio.alignment import global_pairwise_align_protein\n",
        "\n",
        "# Convert strings to Protein objects\n",
        "records = [Protein(s) for s in seqs]\n",
        "\n",
        "# Build TabularMSA object\n",
        "msa = TabularMSA(records)\n",
        "\n",
        "# Print the alignment\n",
        "print(msa)\n",
        "\n",
        "# Extract a column\n",
        "print(\"Column 5:\", msa[:, 5])\n",
        "\n",
        "# Compute a simple majority-rule consensus\n",
        "consensus = msa.consensus()\n",
        "print(\"Consensus:\", consensus)\n",
        "```"
      ],
      "metadata": {
        "id": "gdBu275iw8yi"
      }
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "v1ChqmBXk9RD"
      },
      "source": [
        "## Downloading and reading FASTA files\n",
        "\n",
        "FASTA is a simple, plain-text format commonly used for storing sequence data. Each sequence in a FASTA file includes:\n",
        "- A **header line** starting with `>`, containing metadata such as sequence ID, name, and description.\n",
        "- The **sequence itself** in subsequent lines, represented by single-letter codes for nucleotides or amino acids.\n",
        "\n",
        "#### Example FASTA Format\n",
        "\n",
        ">\\> MCHU - Calmodulin - Human, rabbit, bovine, rat, and chicken\n",
        "MADQLTEEQIAEFKEAFSLFDKDGDGTITTKELGTVMRSLGQNPTEAELQDMINEVDADGNGTID\n",
        "FPEFLTMMARKMKDTDSEEEIREAFRVFDKDGNGYISAAELRHVMTNLGEKLTDEEVDEMIREA\n",
        "DIDGDGQVNYEEFVQMMTAK*\n",
        "\n",
        "In addition to the amino acid sequence, the header line provides metadata about the sequence's origin and description.\n",
        "\n",
        "#### Example sequence\n",
        "\n",
        "For our example, we will use the [CFTR protein](https://www.uniprot.org/uniprot/P13569), associated with cystic fibrosis. Mutations in the CFTR gene are linked to this condition.\n",
        "\n",
        "First, let’s download the `P13569.fasta` file.\n",
        "\n",
        "1. Go to the UniProt link for the CFTR protein: [Cystic fibrosis transmembrane conductance regulator protein | Uniprot](https://www.uniprot.org/uniprot/P13569).\n",
        "2. Click on the **Download** button next to **Tools**.\n",
        "3. Under **Format**, select **FASTA (canonical)**, then click **Generate URL for API**. Finally, click the **Copy** button.\n",
        "\n",
        "You should now have the link: `\"https://rest.uniprot.org/uniprotkb/P13569.fasta\"`. The code for downloading `.fasta` files when you have the URL is shown below."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 18,
      "metadata": {
        "id": "e7_Az_Tmk9RD",
        "outputId": "23a8ce4b-2c85-45a2-a10c-7966de294c37",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Downloaded file to: P13569.fasta\n",
            "\n",
            "Download metadata and statistics:\n",
            "Vary: accept,accept-encoding,x-uniprot-release,x-api-deployment-date, User-Agent\n",
            "Cache-Control: public, max-age=43200\n",
            "x-cache: miss cached\n",
            "Content-Type: text/plain;format=fasta\n",
            "Access-Control-Allow-Credentials: true\n",
            "Access-Control-Expose-Headers: Link, X-Total-Results, X-UniProt-Release, X-UniProt-Release-Date, X-API-Deployment-Date\n",
            "X-API-Deployment-Date: 11-December-2025\n",
            "Strict-Transport-Security: max-age=31536000; includeSubDomains\n",
            "Date: Wed, 07 Jan 2026 06:49:24 GMT\n",
            "Access-Control-Max-Age: 1728000\n",
            "X-UniProt-Release: 2025_04\n",
            "Access-Control-Allow-Origin: *\n",
            "Accept-Ranges: bytes\n",
            "Connection: close\n",
            "Access-Control-Allow-Methods: GET, PUT, POST, DELETE, PATCH, OPTIONS\n",
            "Access-Control-Allow-Headers: DNT,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range,Authorization\n",
            "Content-Length: 1621\n",
            "X-UniProt-Release-Date: 15-October-2025\n",
            "\n",
            "\n"
          ]
        }
      ],
      "source": [
        "from urllib.request import urlretrieve\n",
        "\n",
        "# Specify the destination path for the downloaded file\n",
        "CFTR_FASTA_path = \"P13569.fasta\"\n",
        "\n",
        "# Download the FASTA file from the UniProt URL and save it to the specified path\n",
        "result_location, http_response = urlretrieve(\"https://rest.uniprot.org/uniprotkb/P13569.fasta\", CFTR_FASTA_path)\n",
        "\n",
        "# Confirm the download location\n",
        "print(\"Downloaded file to:\", result_location)\n",
        "\n",
        "# Print metadata and statistics related to the download\n",
        "print(\"\\nDownload metadata and statistics:\")\n",
        "print(http_response)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Pq7fN-xck9RD"
      },
      "source": [
        "The `urlretrieve` function from Python’s `urllib.request` module allows you to download a file from a given URL and save it locally. It takes two main arguments, i.e. `\"https://rest.uniprot.org/uniprotkb/P13569.fasta\"` and `CFTR_FASTA_path`, and returns a tuple containing two elements, i.e. `result_location` and `http_response`.\n",
        "- The `CFTR_FASTA_path` specifies the filename and path where the `.fasta` file will be saved.\n",
        "- The `result_location` confirms the path to the downloaded file.\n",
        "- The `http_response` contains metadata about the download, such as headers, which may include status information and server response details.\n",
        "\n",
        "To open the file and read relevant information from it, we use `SeqIO` module from Biopython."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 19,
      "metadata": {
        "id": "Nh5kr0gfk9RE",
        "outputId": "4a5c7aba-874c-4972-9469-28efed0dcc28",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Total records: 1 \n",
            "\n",
            "First record:\n",
            "ID: sp|P13569|CFTR_HUMAN\n",
            "Name: sp|P13569|CFTR_HUMAN\n",
            "Description: sp|P13569|CFTR_HUMAN Cystic fibrosis transmembrane conductance regulator OS=Homo sapiens OX=9606 GN=CFTR PE=1 SV=3\n",
            "Number of features: 0\n",
            "Seq('MQRSPLEKASVVSKLFFSWTRPILRKGYRQRLELSDIYQIPSVDSADNLSEKLE...TRL')\n",
            "\n",
            "Type of first record: <class 'Bio.SeqRecord.SeqRecord'>\n"
          ]
        }
      ],
      "source": [
        "from Bio import SeqIO\n",
        "\n",
        "# Note that each file can contain multiple records\n",
        "# Initialise an empty list to store the records\n",
        "records = []\n",
        "\n",
        "# Parse the FASTA file and add each record to the list\n",
        "for r in SeqIO.parse(CFTR_FASTA_path, \"fasta\"):\n",
        "    records.append(r)\n",
        "\n",
        "# Print information about the extracted records\n",
        "print(\"Total records:\", len(records), \"\\n\")\n",
        "print(\"First record:\")\n",
        "print(records[0])\n",
        "print(\"\\nType of first record:\", type(records[0]))"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "9mKSmVHYk9RE"
      },
      "source": [
        "- `SeqIO.parse`: Reads the file in FASTA format and returns an iterable of `SeqRecord` objects, one for each sequence in the file.\n",
        "- `records` list: Collects each `SeqRecord` object, allowing us to access the sequences and associated metadata later.\n",
        "- ***Output Details***:\n",
        "  - **Total records**: Displays the count of records in the file, indicating if there are multiple sequences.\n",
        "  - **First record**: Shows the first `SeqRecord` object to provide a sample of the data structure.\n",
        "  - **Type of first record**: Confirms the type of the first item on the list, which should be `SeqRecord`.\n",
        "\n",
        "To get the actual `Seq` object from a `SeqRecord`, we access the `seq` property."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 20,
      "metadata": {
        "id": "XfcJuGNKk9RE",
        "outputId": "df633e80-3604-4718-8020-1b94188e5e23",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Sequence length: 1480 \n",
            "\n",
            "MQRSPLEKASVVSKLFFSWTRPILRKGYRQRLELSDIYQIPSVDSADNLSEKLEREWDRELASKKNPKLINALRRCFFWRFMFYGIFLYLGEVTKAVQPLLLGRIIASYDPDNKEERSIAIYLGIGLCLLFIVRTLLLHPAIFGLHHIGMQMRIAMFSLIYKKTLKLSSRVLDKISIGQLVSLLSNNLNKFDEGLALAHFVWIAPLQVALLMGLIWELLQASAFCGLGFLIVLALFQAGLGRMMMKYRDQRAGKISERLVITSEMIENIQSVKAYCWEEAMEKMIENLRQTELKLTRKAAYVRYFNSSAFFFSGFFVVFLSVLPYALIKGIILRKIFTTISFCIVLRMAVTRQFPWAVQTWYDSLGAINKIQDFLQKQEYKTLEYNLTTTEVVMENVTAFWEEGFGELFEKAKQNNNNRKTSNGDDSLFFSNFSLLGTPVLKDINFKIERGQLLAVAGSTGAGKTSLLMVIMGELEPSEGKIKHSGRISFCSQFSWIMPGTIKENIIFGVSYDEYRYRSVIKACQLEEDISKFAEKDNIVLGEGGITLSGGQRARISLARAVYKDADLYLLDSPFGYLDVLTEKEIFESCVCKLMANKTRILVTSKMEHLKKADKILILHEGSSYFYGTFSELQNLQPDFSSKLMGCDSFDQFSAERRNSILTETLHRFSLEGDAPVSWTETKKQSFKQTGEFGEKRKNSILNPINSIRKFSIVQKTPLQMNGIEEDSDEPLERRLSLVPDSEQGEAILPRISVISTGPTLQARRRQSVLNLMTHSVNQGQNIHRKTTASTRKVSLAPQANLTELDIYSRRLSQETGLEISEEINEEDLKECFFDDMESIPAVTTWNTYLRYITVHKSLIFVLIWCLVIFLAEVAASLVVLWLLGNTPLQDKGNSTHSRNNSYAVIITSTSSYYVFYIYVGVADTLLAMGFFRGLPLVHTLITVSKILHHKMLHSVLQAPMSTLNTLKAGGILNRFSKDIAILDDLLPLTIFDFIQLLLIVIGAIAVVAVLQPYIFVATVPVIVAFIMLRAYFLQTSQQLKQLESEGRSPIFTHLVTSLKGLWTLRAFGRQPYFETLFHKALNLHTANWFLYLSTLRWFQMRIEMIFVIFFIAVTFISILTTGEGEGRVGIILTLAMNIMSTLQWAVNSSIDVDSLMRSVSRVFKFIDMPTEGKPTKSTKPYKNGQLSKVMIIENSHVKKDDIWPSGGQMTVKDLTAKYTEGGNAILENISFSISPGQRVGLLGRTGSGKSTLLSAFLRLLNTEGEIQIDGVSWDSITLQQWRKAFGVIPQKVFIFSGTFRKNLDPYEQWSDQEIWKVADEVGLRSVIEQFPGKLDFVLVDGGCVLSHGHKQLMCLARSVLSKAKILLLDEPSAHLDPVTYQIIRRTLKQAFADCTVILCEHRIEAMLECQQFLVIEENKVRQYDSIQKLLNERSLFRQAISPSDRVKLFPHRNSSKCKSKPQIAALKEETEEEVQDTRL\n"
          ]
        }
      ],
      "source": [
        "# Access the first record from the list of records\n",
        "record = records[0]\n",
        "\n",
        "# Print the length of the sequence\n",
        "print(\"Sequence length:\", len(record.seq), \"\\n\")\n",
        "\n",
        "# Print the sequence itself\n",
        "print(record.seq)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "collapsed": false,
        "jupyter": {
          "outputs_hidden": false
        },
        "id": "Hs_Azsllk9RE"
      },
      "source": [
        "## Folding\n",
        "\n",
        "We have covered the initial steps of the ***central dogma of molecular biology***: transcribing DNA into RNA and then translating it into a protein sequence. The final step, i.e. converting a sequence into a protein structure, is known as **protein folding**. This process is considerably more complex and computationally expensive to predict.\n",
        "\n",
        "One method that has recently gained prominence is **[AlphaFold](https://www.nature.com/articles/s41586-021-03819-2)**. It predicts protein structures based on known folding patterns and sequence alignments, using deep learning techniques to achieve remarkable accuracy.\n",
        "The **[2024 Nobel Prize in Chemistry](https://www.nature.com/articles/d41586-024-03214-7)** was awarded to John Jumper and Demis Hassabis of Google DeepMind for developing this AI tool, and to David Baker for his work on computational protein design. Their contributions have revolutionised biology and hold the potential to transform drug discovery.\n",
        "\n",
        "You can make your own predictions by using [the official AlphaFold notebook](https://colab.research.google.com/github/deepmind/alphafold/blob/main/notebooks/AlphaFold.ipynb#scrollTo=rowN0bVYLe9n) on Google Colab."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "JH6Osl6sk9RE"
      },
      "source": [
        "## Discussion\n",
        "\n",
        "In this notebook, we explored fundamental ways to construct and manipulate sequences, including transcription and translation. We examined how **Biopython** facillitates *sequence alignment*, both pairwise and multiple sequence alignment (MSA), and demonstrated how to *download and read FASTA files* to access sequence records and metadata.\n",
        "\n",
        "This notebook provides only an introduction to the extensive tools and integrations available within the **Biopython** library, which supports a wide range of formats and methods for storing and processing sequence and alignment data.\n",
        "\n",
        "Please note, this notebook uses very simple sequences. In real-world applications, sequences are often much longer and more complex. We encourage you to revisit this notebook with longer sequences and to experiment with the code to gain a deeper understanding.\n",
        "\n",
        "You can now move on to the exercise notebook which will allow you to use the code introduced here in more practical scenarios.\n",
        "\n",
        "If you want to learn more there are some extra external resources linked at the beginning of this notebook. You can click [here](#Contents) to go back to the top."
      ]
    }
  ],
  "metadata": {
    "celltoolbar": "Edit Metadata",
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.7"
    },
    "pycharm": {
      "stem_cell": {
        "cell_type": "raw",
        "metadata": {
          "collapsed": false
        },
        "source": []
      }
    },
    "colab": {
      "provenance": []
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}